en/docs/extended_topics/user_op.md

```diff
@@ -1,414 +1,816 @@
+  1. `Input("in") ` means that it has an input named "in".
+  2. `Output("out")` means that it has an output named "out".
+  3. `SetTensorDescInferFn` is used to set the shape and data type of the derivation function which describe the relationship between the input of this operator and shape and type of the output of this operator. In the above code, the shape and data type of the output is consistent with input.
+
+### Implementation and Registration of CPU Kernel 
+
+We implemented the CPU version of the kernel in `myrelu_cpu_kernel.cpp` and registered it：
 
+```cpp
+#include "oneflow/core/framework/framework.h"
+
+namespace oneflow {
+
+namespace {
+
+template <typename T>
+void MyRelu(DeviceCtx *ctx, const int64_t n, const T *x, T *y) {
+  T zero = (T)(0);
+  for (int64_t i = 0; i != n; ++i) {
+    y[i] = std::max(x[i], zero);
+  }
 }
+
+template <DeviceType device_type, typename T>
+class ReluKernel final : public user_op::OpKernel {
+public:
+  ReluKernel() = default;
+  ~ReluKernel() = default;
+
+private:
+  void Compute(user_op::KernelComputeContext *ctx) const override {
+    const user_op::Tensor *in_tensor = ctx->Tensor4ArgNameAndIndex("in", 0);
+    user_op::Tensor *out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    MyRelu<T>(ctx->device_ctx(),
+           in_tensor->shape().elem_cnt(),
+           in_tensor->dptr<T>(), 
+           out_tensor->mut_dptr<T>());
+  }
+  bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
+};
+
+#define REGISTER_RELU_KERNEL(device, dtype)          \
+  REGISTER_USER_KERNEL("myrelu")                     \
+      .SetCreateFn<ReluKernel<device, dtype>>()      \
+      .SetIsMatchedHob(                              \
+          (user_op::HobDeviceType() == device) &     \
+          (user_op::HobDataType("out", 0)            \
+            == GetDataType<dtype>::value));
+
+REGISTER_RELU_KERNEL(DeviceType::kCPU, float)
+REGISTER_RELU_KERNEL(DeviceType::kCPU, double)
+} // namespace
+
+} // namespace oneflow
 ```
 
-分析下上述代码：
+To implement the kernel in OneFlow, you must define a class which inherits from `oneflow::user_op::OpKernel` and rewrite the virtual functions of it.
 
-首先是include了`oneflow/core/framework/framework.h`。`framework.h`中包括了我们创建一个Op需要的所有头文件，所以我们只需要include这一个头文件就行。
+In the above code, we rewrite `Compute` and `AlwaysComputeWhenAllOutputsEmpty` and their respective meanings are:
 
-其次，为了简化类型名，我们在`namespace oneflow`下书写注册Op定义的相关代码。
+* `Compute` must be rewritten to implement the specific operating logic.
 
-最后，我们来看下注册的代码。`Relu`  op只有一个输入`in`，也只有一个输出`out`，其推导out tensor的形状的`ShapeInferFn`是一个lambda函数，参数为`InferContext`，返回值为`Maybe<void>`类型对象。
+* `AlwaysComputeWhenAllOutputsEmpty` must be rewritten to return `false` for the most of op. For the some op that needs to maintain state internally. Therefore needs to be computed by calling the kernel even if the output is null and return `true`.
```

Comment:
`AlwaysComputeWhenAllOutputsEmpty` must be rewritten to return `false` in most cases.

Reply:
For very few ops that need to maintain state internally, and therefore need to call the kernel for calculation even if the output is empty, it should return true